### ISSUE TITLE
DnnImageHandler does not set sensible browser cache headers

### ISSUE THREAD

**Issue by OllyHodgson** [dnnsoftware/Dnn.Platform#2150]
## Description 

DnnImageHandler does not set a max-age or expires HTTP header. This means the content will not be effectively cached by the browser.

## Steps to reproduce

I've got a custom OpenContent module, which I'm using to serve news articles. Sometimes my users upload astoundingly large images, so to work around this I've used `DnnImageHandler.ashx` to serve something appropriate. e.g. `/DnnImageHandler.ashx?mode=file&w=400&file=/Portals/0/OpenContent/Files/385/mma-mobile-bg.jpg`

## Current result

When I run this through the webpagetest.org, it reports:

> **Leverage browser caching of static assets: 50/100**
> FAILED - (**No max-age or expires**) - https://url-to-my-site/DnnImageHandler.ashx?mode=file&w=400&file=/Portals/0/OpenContent/Files/385/mma-mobile-bg.jpg
> FAILED - (**No max-age or expires**) - https://url-to-my-site/DnnImageHandler.ashx?mode=file&w=400&file=/Portals/0/OpenContent/Files/388/tyler-gebhart-488825-unsplash.jpg

## Expected result

DnnImageHandler.ashx sets an appropriate [`Cache-Control: max-age` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Expiration) (e.g. `Cache-Control: max-age=86400`) along with an appropriate [`ETag` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) (e.g. an ASCII hash of the image) or [Last-Modified header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified).

*Ideally* I can disable this and/or customise the max-age with a querystring argument (setting it to 0 seconds would effectively disable it).

## Affected version

I've tested with 9.2.0.

## Affected browser

Any browser.

**Comment by OllyHodgson** [comment]
*TL;DR:* Fix or remove the nonsensical `Last-Modified` header. Set the `Expires` header to the far future.

It turns out DnnImageHandler *does* return `Cache-Control`, `Expires`, `Last-Modified` and `ETag` headers (I've stripped out the irrelevant ones):

```
Cache-Control: public
Date: Wed, 11 Jul 2018 09:09:52 GMT
Expires: Wed, 11 Jul 2018 09:19:52 GMT
Last-Modified: Wed, 11 Jul 2018 09:09:52 GMT
ETag: 2B521CFA9E685E2147DB92759B354FA2745901A1
```

The `ETag` appears to be correct, and stays the same on subsequent requests. However, the `Last-Modified` header contradicts it: it's set to the same time as the request, rather than when the image was actually generated. This could be a bit confusing for the browser - the image is unchanged, but it was last modified a moment ago? What do I do? Assuming the generated images are being cached on the server, perhaps use their created date this header? Even easier would be to remove the header altogether and rely on ETag.

Secondly, the expires header is set only 10 minutes into the future. Unless this is being used for very rapidly changing content, this ought to be set a lot longer so the browser doesn't try to reload the image if the user comes back to the page 20 minutes later. If [Daniel Mettler's blog post](http://www.dnnsoftware.com/community-blog/cid/155397/get-google-pagespeed-points-with-a-quick-webconfig-setting) is anything to go by, setting it to 30 days (2592000 seconds) would be better.

Finally, it appears that the caching behaviour can be disabled (e.g. for rapidly changing content) using the existing `NoCache` querystring parameter.

**Comment by sleupold** [comment]
@OllyHodgson are you able to submit a pull request for this issue?

**Comment by bdukes** [comment]
You can configure the `Expires` header using an app setting in the `web.config`, like this (see [`ReadSettings` method](https://github.com/dnnsoftware/Dnn.Platform/blob/dbf0e4e33692155eb241559807234caed76ea2be/DNN%20Platform/Library/Services/GeneratedImage/DnnImageHandler.cs#L483-L533)):
```
<add key="DnnImageHandler" value="clientcacheexpiration=2592000" />
```

**Comment by bdukes** [comment]
And [here's where it sets `Last-Modified`](https://github.com/dnnsoftware/Dnn.Platform/blob/dbf0e4e33692155eb241559807234caed76ea2be/DNN%20Platform/Library/Services/GeneratedImage/ImageHandlerInternal.cs#L268), not sure if that should just be removed, or set to a different value…
